File: Languages/Python/math.py
print("Hello, World!")

import re
import json
import yaml
import os

def solve_equation(equation):
    # Remove the "= ?" part and evaluate the expression
    equation = equation.replace("= ?", "").strip()
    
    # Use eval to solve the math equation (safe in this controlled environment)
    try:
        result = eval(equation)
        return f"{equation} = {result}"
    except:
        return f"{equation} = Error"

def process_txt_file(filepath):
    with open(filepath, 'r') as f:
        lines = f.readlines()
    
    results = []
    for line in lines:
        line = line.strip()
        if line and '=' in line:
            results.append(solve_equation(line))
    
    return results

def process_md_file(filepath):
    with open(filepath, 'r') as f:
        content = f.read()
    
    # Extract equations from markdown
    lines = content.split('\n')
    results = []
    
    for line in lines:
        line = line.strip()
        if line.startswith('- ') and '=' in line:
            equation = line[2:]  # Remove "- "
            results.append(solve_equation(equation))
    
    return results

def process_json_file(filepath):
    with open(filepath, 'r') as f:
        data = json.load(f)
    
    results = []
    for item in data['equations']:
        equation = f"{item['expression']} = ?"
        results.append(solve_equation(equation))
    
    return results

def process_yaml_file(filepath):
    with open(filepath, 'r') as f:
        data = yaml.safe_load(f)
    
    results = []
    for item in data['equations']:
        equation = f"{item['expression']} = ?"
        results.append(solve_equation(equation))
    
    return results

def process_raw_file(filepath):
    with open(filepath, 'r') as f:
        lines = f.readlines()
    
    results = []
    for line in lines:
        line = line.strip()
        if line and '=' in line:
            results.append(solve_equation(line))
    
    return results

def main():
    test_data_dir = "../test_data"
    
    # Process all files
    files = [
        ("math_equations.txt", process_txt_file),
        ("math_equations.md", process_md_file),
        ("math_equations.json", process_json_file),
        ("math_equations.yaml", process_yaml_file),
        ("math_equations", process_raw_file)
    ]
    
    all_results = []
    
    for filename, processor in files:
        filepath = os.path.join(test_data_dir, filename)
        if os.path.exists(filepath):
            print(f"\nProcessing {filename}:")
            results = processor(filepath)
            all_results.extend(results)
            for result in results:
                print(result)
    
    print(f"\nTotal equations solved: {len(all_results)}")

if __name__ == "__main__":
    main()

File: Languages/Haskell/Math.hs
module Main where

import System.IO
import Data.List
import Data.Char
import Text.Read

-- Hello, World!
hello :: IO ()
hello = putStrLn "Hello, World!"

-- Simple expression evaluator
evaluateExpression :: String -> Double
evaluateExpression expr = 
    let parts = words expr
    in if length parts < 3 
       then 0.0
       else case (readMaybe (parts !! 0) :: Maybe Double,
                  readMaybe (parts !! 2) :: Maybe Double) of
                (Just a, Just b) -> 
                    case parts !! 1 of
                        "+" -> a + b
                        "-" -> a - b
                        "*" -> a * b
                        "/" -> if b /= 0.0 then a / b else 0.0
                        _ -> 0.0
                _ -> 0.0

solveEquation :: String -> IO ()
solveEquation line = do
    let equation = takeWhile (/= '=') line
    let trimmedEquation = dropWhile isSpace equation
    let result = evaluateExpression trimmedEquation
    putStrLn $ trimmedEquation ++ " = " ++ show (result :: Double)

processFile :: FilePath -> IO ()
processFile filename = do
    exists <- doesFileExist filename
    if not exists
        then putStrLn $ "Could not open file: " ++ filename
        else do
            contents <- readFile filename
            let lines' = lines contents
            mapM_ processLine lines'
  where
    processLine line
        | null (dropWhile isSpace line) = return ()
        | head (dropWhile isSpace line) == '#' = return ()
        | otherwise = do
            let start = if take 2 line == "- " then drop 2 line else line
            if '=' `elem` start
                then solveEquation start
                else return ()

main :: IO ()
main = do
    hello
    putStrLn "\nProcessing math equations..."
    
    processFile "../test_data/math_equations.txt"
    processFile "../test_data/math_equations.md"
    processFile "../test_data/math_equations.json"
    processFile "../test_data/math_equations.yaml"
    processFile "../test_data/math_equations"
